test/e2e/external_oidc_test.go

```diff
@@ -0,0 +1,449 @@
+	defer cancel()
+
+	tc, err := newTestClient(t)
+	require.NoError(t, err)
+
+	// ===============================
+	// Check ExternalOIDC feature gate
+	// ===============================
+
+	featureGates, err := tc.configClient.ConfigV1().FeatureGates().Get(testCtx, "cluster", metav1.GetOptions{})
+	require.NoError(tc.t, err)
+
+	if len(featureGates.Status.FeatureGates) != 1 {
+		// skip test if there are multiple feature gate versions (i.e. ongoing upgrade)
+		tc.t.Skipf("multiple feature gate versions detected")
+	} else {
+		for _, fgDisabled := range featureGates.Status.FeatureGates[0].Disabled {
+			if fgDisabled.Name == featureGateExternalOIDC {
+				tc.t.Skipf("feature gate '%s' disabled", featureGateExternalOIDC)
+			}
+		}
+	}
+
+	// ==============
+	// Setup Keycloak
+	// ==============
+
+	var kcClient *test.KeycloakClient
+	if keycloakURL := os.Getenv("E2E_KEYCLOAK_URL"); len(keycloakURL) > 0 {
+		t.Logf("will use existing keycloak deployment at URL: %s", keycloakURL)
+		kcClient = tc.setupKeycloakClient(testCtx, keycloakURL)
+
+	} else {
+		t.Logf("no existing keycloak deployment found; will create new")
+		var cleanups []func()
+		kcClient, cleanups = tc.setupExternalOIDCWithKeycloak(testCtx)
+		defer test.IDPCleanupWrapper(func() {
+			for _, c := range cleanups {
+				c()
+			}
+		})()
+		t.Logf("keycloak Admin URL: %s", kcClient.AdminURL())
+	}
+
+	kcAdminClient, err := kcClient.GetClientByClientID(oidcClientId)
+	require.NoError(t, err)
+	require.NotEmpty(t, kcAdminClient)
+
+	group := names.SimpleNameGenerator.GenerateName("e2e-keycloak-group-")
+	err = kcClient.CreateGroup(group)
+	require.NoError(t, err)
+
+	user := names.SimpleNameGenerator.GenerateName("e2e-keycloak-user-")
+	password := "password"
+	err = kcClient.CreateUser(user, password, []string{group})
+	require.NoError(t, err)
+
+	httpClient := &http.Client{Transport: &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	}}
+
+	formData := url.Values{
+		"grant_type": []string{"password"},
+		"client_id":  []string{oidcClientId},
+		"scope":      []string{"openid"},
+		"username":   []string{user},
+		"password":   []string{password},
+	}
+
+	resp, err := httpClient.PostForm(kcClient.TokenURL(), formData)
+	require.NoError(t, err)
+	defer resp.Body.Close()
+	require.Equal(t, http.StatusOK, resp.StatusCode)
+
+	data, err := io.ReadAll(resp.Body)
+	require.NoError(t, err)
+	require.NotEmpty(t, data)
+
+	var authResponse oidcAuthResponse
+	err = json.Unmarshal(data, &authResponse)
+	require.NoError(t, err)
+	require.NotEmpty(t, authResponse.AccessToken)
+	require.NotEmpty(t, authResponse.IdToken)
+
+	// ==========================================
+	// Test authentication via the kube-apiserver
+	// ==========================================
+
+	// use the OIDC id_token to do a self subject access review
+	kasURL := fmt.Sprintf("%s/apis/authorization.k8s.io/v1/selfsubjectaccessreviews", tc.kubeConfig.Host)
```

Comment:
https://kubernetes.io/docs/reference/kubernetes-api/authentication-resources/self-subject-review-v1/ actually returns the username/groups of the authenticated user making the request. If I'm understanding the test, we want to verify that the provided token is authenticated properly. If the test is only checking that the requester is _authorized_ to make a certain request, that's more likely to give a false-positive result.